File: ShowMIDIPackets/ShowMIDIPackets/MessageParser.m
#import "MessageParser.h"

typedef enum _SMSystemCommonMessageType {
    SMSystemCommonMessageTypeTimeCodeQuarterFrame = 0xF1,
    SMSystemCommonMessageTypeSongPositionPointer = 0xF2,
    SMSystemCommonMessageTypeSongSelect = 0xF3,
    SMSystemCommonMessageTypeTuneRequest = 0xF6
} SMSystemCommonMessageType;

typedef enum _SMSystemRealTimeMessageType {
    SMSystemRealTimeMessageTypeClock = 0xF8,
    SMSystemRealTimeMessageTypeStart = 0xFA,
    SMSystemRealTimeMessageTypeContinue = 0xFB,
    SMSystemRealTimeMessageTypeStop = 0xFC,
    SMSystemRealTimeMessageTypeActiveSense = 0xFE,
    SMSystemRealTimeMessageTypeReset = 0xFF
} SMSystemRealTimeMessageType;

@interface MessageParser ()

@property (nonatomic) NSMutableData *readingSysExData;

@end

@implementation MessageParser

- (void)takePacketList:(const MIDIPacketList *)packetList
{
    if (!packetList || !packetList->numPackets) {
        return;
    }

    NSMutableArray *messages = [NSMutableArray array];

    UInt32 packetCount = packetList->numPackets;
    const MIDIPacket *packet = packetList->packet;
    while (packetCount--) {
        NSArray *messagesForPacket = [self messagesForPacket:packet];
        if (messagesForPacket) {
            [messages addObjectsFromArray:messagesForPacket];
        }

        packet = MIDIPacketNext(packet);
    }

    if (messages.count) {
        [self.delegate parser:self didReadMessages:messages];
    }
}

#pragma mark Private

- (NSArray *)messagesForPacket:(const MIDIPacket *)packet
{
    // Split this packet into separate MIDI messages    
    NSMutableArray *messages = nil;
    const Byte *data;
    UInt16 length;
    Byte byte;
    Byte pendingMessageStatus;
    Byte pendingData[2];
    UInt16 pendingDataIndex, pendingDataLength;
    NSMutableData* readingInvalidData = nil;
    
    pendingMessageStatus = 0;
    pendingDataIndex = pendingDataLength = 0;

    data = packet->data;
    length = packet->length;
    while (length--) {
        NSString *message = nil;
        BOOL byteIsInvalid = NO;
        
        byte = *data++;
    
        if (byte >= 0xF8) {
            // Real Time message    
            switch (byte) {
                case SMSystemRealTimeMessageTypeClock:
                case SMSystemRealTimeMessageTypeStart:
                case SMSystemRealTimeMessageTypeContinue:
                case SMSystemRealTimeMessageTypeStop:
                case SMSystemRealTimeMessageTypeActiveSense:
                case SMSystemRealTimeMessageTypeReset:
                    message = [NSString stringWithFormat:@"SystemRealTime %02X", byte];
                    break;
        
                default:
                    // Byte is invalid
                    byteIsInvalid = YES;
                    break;
            }
        } else {
            if (byte < 0x80) {
                if (self.readingSysExData) {
                    [self.readingSysExData appendBytes:&byte length:1];
                } else if (pendingDataIndex < pendingDataLength) {
                    pendingData[pendingDataIndex] = byte;
                    pendingDataIndex++;

                    if (pendingDataIndex == pendingDataLength) {
                        // This message is now done--send it
                        if (pendingMessageStatus >= 0xF0) {
                            message = [NSString stringWithFormat:@"SystemCommon %02X %@", pendingMessageStatus, [self formatData:pendingData withLength:pendingDataLength]];
                        } else {
                            message = [NSString stringWithFormat:@"Voice %02X %@", pendingMessageStatus, [self formatData:pendingData withLength:pendingDataLength]];
                        }

                        pendingDataLength = 0;
                    }                    
                } else {
                    // Skip this byte -- it is invalid
                    byteIsInvalid = YES;
                }
            } else {
                if (self.readingSysExData) {
                    message = [self finishSysExMessageWithValidEnd:(byte == 0xF7)];
                }

                pendingMessageStatus = byte;
                pendingDataLength = 0;
                pendingDataIndex = 0;
                
                switch (byte & 0xF0) {            
                    case 0x80:	// Note off
                    case 0x90:	// Note on
                    case 0xA0:	// Aftertouch
                    case 0xB0:	// Controller
                    case 0xE0:	// Pitch wheel
                        pendingDataLength = 2;
                        break;
    
                    case 0xC0:	// Program change
                    case 0xD0:	// Channel pressure
                        pendingDataLength = 1;
                        break;
                    
                    case 0xF0: {
                        // System common message
                        switch (byte) {
                            case 0xF0:
                                // System exclusive
                                self.readingSysExData = [[NSMutableData alloc] init];
                                break;
                                
                            case 0xF7:
                                // System exclusive ends--already handled above.
                                // But if this is showing up outside of sysex, it's invalid.
                                if (!message) {
                                    byteIsInvalid = YES;
                                }
                                break;
                            
                            case SMSystemCommonMessageTypeTimeCodeQuarterFrame:
                            case SMSystemCommonMessageTypeSongSelect:
                                pendingDataLength = 1;
                                break;
    
                            case SMSystemCommonMessageTypeSongPositionPointer:
                                pendingDataLength = 2;
                                break;
    
                            case SMSystemCommonMessageTypeTuneRequest:
                                message = [NSString stringWithFormat:@"SystemCommon %02X", byte];
                                break;
                            
                            default:
                                // Invalid message
                                byteIsInvalid = YES;
                                break;
                        }                 
                        break;
                    }
                
                    default:
                        // This can't happen, but handle it anyway
                        byteIsInvalid = YES;
                        break;
                }
            }
        }

        if (byteIsInvalid) {
            if (!readingInvalidData) {
                readingInvalidData = [NSMutableData data];
            }
            [readingInvalidData appendBytes:&byte length:1];
        }

        if (readingInvalidData && (!byteIsInvalid || length == 0)) {
            // We hit the end of a stretch of invalid data.
            message = [NSString stringWithFormat:@"Invalid %@", [self formatData:readingInvalidData]];
            readingInvalidData = nil;
        }

        if (message) {
            if (!messages) {
                messages = [NSMutableArray arrayWithObject:message];
            } else {
                [messages addObject:message];
            }
        }
    }

    return messages;
}

- (NSString *)finishSysExMessageWithValidEnd:(BOOL)isEndValid
{
    NSString *message = nil;

    if (self.readingSysExData) {
        message = [NSString stringWithFormat:@"SysEx %@", [self formatData:self.readingSysExData]];
        self.readingSysExData = nil;
    }

    return message;
}

- (NSString *)formatData:(Byte *)data withLength:(UInt16)length
{
    if (!data) {
        return @"NULL";
    }

    NSMutableString *str = [NSMutableString string];
    for (UInt16 i = 0; i < length; i++) {
        [str appendFormat:@"%02X ", data[i]];
    }
    return str;
}

- (NSString *)formatData:(NSData *)data
{
    if (!data) {
        return @"(nil)";
    }

    Byte* bytes = (Byte *)data.bytes;
    if (!data) {
        return @"NSData with NULL bytes";
    }

    NSMutableString *str = [NSMutableString string];
    for (NSUInteger i = 0; i < data.length; i++) {
        [str appendFormat:@"%02X ", bytes[i]];
    }
    return str;
}

@end

File: Frameworks/SnoizeMIDI/MessageParser.swift
/*
 Copyright (c) 2001-2021, Kurt Revis.  All rights reserved.

 This source code is licensed under the BSD-style license found in the
 LICENSE file in the root directory of this source tree.
 */

import Foundation

protocol MessageParserDelegate: AnyObject {

    func parserDidReadMessages(_ parser: MessageParser, messages: [Message])
    func parserIsReadingSysEx(_ parser: MessageParser, length: Int)
    func parserFinishedReadingSysEx(_ parser: MessageParser, message: SystemExclusiveMessage)

}

public class MessageParser {

    deinit {
        sysExTimeOutTimer?.invalidate()
    }

    weak var delegate: MessageParserDelegate?
    public weak var originatingEndpoint: Endpoint?
    public var sysExTimeOut: TimeInterval = 1.0   // seconds
    public var ignoresInvalidData = false

    public func takePacketList(_ packetListPtr: UnsafePointer<MIDIPacketList>) {
        var messages: [Message] = []

        if #available(macOS 10.15, iOS 13.0, *) {
            messages = packetListPtr.unsafeSequence().flatMap { messagesForPacket($0) }
        }
        else {
            // Fallback on earlier versions
            SMPacketListApply(packetListPtr) {
                messages.append(contentsOf: messagesForPacket($0))
            }
        }

        if !messages.isEmpty {
            delegate?.parserDidReadMessages(self, messages: messages)
        }

        if readingSysExData != nil {
            if sysExTimeOutTimer == nil {
                // Create a timer which will fire after we have received no sysex data for a while.
                // This takes care of interruption in the data (devices being turned off or unplugged) as well as
                // ill-behaved devices which don't terminate their sysex messages with 0xF7.
                sysExTimeOutTimer = Timer.scheduledTimer(timeInterval: sysExTimeOut, target: self, selector: #selector(sysExTimedOut(_:)), userInfo: nil, repeats: false)
            }
            else {
                // We already have a timer, so just bump its fire date to later.
                sysExTimeOutTimer?.fireDate = Date(timeIntervalSinceNow: sysExTimeOut)
            }
        }
        else {
            // Not reading sysex, so if we have a timeout pending, forget about it
            sysExTimeOutTimer?.invalidate()
            sysExTimeOutTimer = nil
        }
    }

    @discardableResult public func cancelReceivingSysExMessage() -> Bool {
        // Returns YES if it successfully cancels a sysex message which is being received, and NO otherwise.
        if readingSysExData != nil {
            readingSysExData = nil
            return true
        }
        else {
            return false
        }
    }

    // MARK: Private

    private var readingSysExData: Data?
    private var startSysExTimeStamp: MIDITimeStamp = 0
    private var sysExTimeOutTimer: Timer?

    private struct PendingMessage {
        var status: UInt8 = 0
        var data: [UInt8] = []
        var expectedCount: Int = 0
    }

    private func messagesForPacket(_ packetPtr: UnsafePointer<MIDIPacket>) -> [Message] {
        // Split this packet into separate MIDI messages.

        let packetDataCount = Int(packetPtr.pointee.length)
        guard packetDataCount > 0 else { return [] }
        let timeStamp = packetPtr.pointee.timeStamp

        var pendingMessage = PendingMessage()
        var readingInvalidData: Data?

        // Safely getting to the packet data is more difficult than it should be.
        // Can't use withUnsafePointer(to: packetPtr.pointee.data.0) since that crashes with ASAN on.
        // (Accessing `pointee` appears to be trying to copy 256 bytes of data, which may be more than
        // is really accessible.)
        // Can't use withUnsafeBytes(of: packetPtr.pointee.data) since that limits to the 256 bytes
        // in the tuple in the struct. There may be more.
        // Do it the hard way instead.
        let rawPacketDataPtr = UnsafeRawBufferPointer(start: UnsafeRawPointer(packetPtr) + MemoryLayout.offset(of: \MIDIPacket.data)!, count: packetDataCount)
        return rawPacketDataPtr.enumerated().flatMap { (byteIndex, byte) -> [Message] in
            var messages: [Message] = []
            var byteIsInvalid = false

            var maybeMessage: Message?
            if byte >= 0xF8 {
                // Real Time message, always one byte, may be interspersed anywhere in other messages (including SysEx)
                (maybeMessage, byteIsInvalid) = parseRealTimeMessage(byte, timeStamp)
            }
            else if byte < 0x80 {
                // Message data byte, goes into pendingMessage, might complete a multibyte message
                (maybeMessage, byteIsInvalid) = parseMessageData(byte, timeStamp, &pendingMessage)
            }
            else if byte == 0xF7 {
                // Explicit end of a SysEx message
                if let sysExMessage = finishSysExMessage(validEnd: true) {
                    messages.append(sysExMessage)
                }
                else {
                    // SysEx end byte with no SysEx message being read. Technically invalid.
                    byteIsInvalid = true
                }
            }
            else {
                // Start of a non-real-time message.
                // This terminates the current SysEx message, if any
                if let sysExMessage = finishSysExMessage(validEnd: false) {
                    messages.append(sysExMessage)
                }
                (maybeMessage, byteIsInvalid) = parseMessageStart(byte, timeStamp, &pendingMessage)
            }

            if let message = maybeMessage {
                messages.append(message)
            }

            if !ignoresInvalidData,
               let invalidMessage = parsePotentiallyInvalidByte(byte, timeStamp, byteIsInvalid, (byteIndex == packetDataCount - 1), &readingInvalidData) {
                messages.append(invalidMessage)
            }

            messages.forEach { $0.originatingEndpoint = originatingEndpoint }
            return messages
        }
    }

    private func parseRealTimeMessage(_ byte: UInt8, _ timeStamp: MIDITimeStamp) -> (Message?, Bool) {
        // Real Time message
        if let messageType = SystemRealTimeMessage.MessageType(rawValue: byte) {
            return (SystemRealTimeMessage(timeStamp: timeStamp, type: messageType), false)
        }
        else {
            return (nil, true)
        }
    }

    private func parseMessageData(_ byte: UInt8, _ timeStamp: MIDITimeStamp, _ pendingMessage: inout PendingMessage) -> (Message?, Bool) {
        // N.B. Be careful about performance here. Doing something like this will be slow,
        //      probably N^2 with the number of sysex bytes read:
        //
        //     if let data = readingSysExData {
        //         readingSysExData?.append(byte)
        //         ...
        //
        // `readingSysExData` is a `Data` which is a value object, and thus `if let` or `if var` always copy
        // that value object. (See: https://github.com/swiftlang/swift-evolution/blob/main/proposals/0345-if-let-shorthand.md
        // which says "optional binding conditions always make a copy of the value"; the hoped-for fix in
        // https://forums.swift.org/t/a-roadmap-for-improving-swift-performance-predictability-arc-improvements-and-ownership-control/54206/268
        // has not yet landed.)
        //
        // The perf problem appears when we append the byte, because that's when COW actually needs to copy the original bytes.

        if readingSysExData != nil {
            readingSysExData!.append(byte)

            // Tell the delegate we're still reading, every 256 bytes
            let sysExMessageDataCount = 1 /* for 0xF0 */ + readingSysExData!.count
            if sysExMessageDataCount % 256 == 0 {
                delegate?.parserIsReadingSysEx(self, length: sysExMessageDataCount)
            }
        }
        else if pendingMessage.data.count < pendingMessage.expectedCount {
            pendingMessage.data.append(byte)

            if pendingMessage.data.count == pendingMessage.expectedCount {
                // This message is now done
                if let status = SystemCommonMessage.Status(rawValue: pendingMessage.status) {
                    return (SystemCommonMessage(timeStamp: timeStamp, status: status, data: pendingMessage.data), false)
                }
                else {
                    return (VoiceMessage(timeStamp: timeStamp, statusByte: pendingMessage.status, data: pendingMessage.data), false)
                }
            }
        }
        else {
            // Skip this byte -- it is invalid
            return (nil, true)
        }

        // No message yet, but this byte was valid
        return (nil, false)
    }

    private func parseMessageStart(_ byte: UInt8, _ timeStamp: MIDITimeStamp, _ pendingMessage: inout PendingMessage) -> (Message?, Bool) {
        var message: Message?
        var byteIsInvalid = false

        pendingMessage.status = byte
        pendingMessage.data = []
        pendingMessage.expectedCount = 0

        switch byte & 0xF0 {
        case 0x80,    // Note off
             0x90,    // Note on
             0xA0,    // Aftertouch
             0xB0,    // Controller
             0xE0:    // Pitch wheel
            pendingMessage.expectedCount = 2

        case 0xC0,    // Program change
             0xD0:    // Channel pressure
            pendingMessage.expectedCount = 1

        case 0xF0:
            // System common message
            if byte == 0xF0 {
                // System exclusive start
                readingSysExData = Data()
                startSysExTimeStamp = timeStamp
                delegate?.parserIsReadingSysEx(self, length: 1)
            }
            else if let systemCommonMessageStatus = SystemCommonMessage.Status(rawValue: byte) {
                let dataLength = systemCommonMessageStatus.otherDataLength
                if dataLength > 0 {
                    pendingMessage.expectedCount = dataLength
                }
                else {
                    message = SystemCommonMessage(timeStamp: timeStamp, status: systemCommonMessageStatus, data: [])
                }
            }
            else {
                // Invalid message
                byteIsInvalid = true
            }

        default:
            // This can't happen, but handle it anyway
            byteIsInvalid = true
        }

        return (message, byteIsInvalid)
    }

    private func finishSysExMessage(validEnd: Bool) -> SystemExclusiveMessage? {
        // NOTE: If we want, we could refuse sysex messages that don't end in 0xF7.
        // The MIDI spec says that messages should end with this byte, but apparently that is not always the case in practice.
        guard let data = readingSysExData else { return nil }
        readingSysExData = nil

        let message = SystemExclusiveMessage(timeStamp: startSysExTimeStamp, data: data)
        message.originatingEndpoint = originatingEndpoint
        message.wasReceivedWithEOX = validEnd
        delegate?.parserFinishedReadingSysEx(self, message: message)

        return message
    }

    private func parsePotentiallyInvalidByte(_ byte: UInt8, _ timeStamp: MIDITimeStamp, _ byteIsInvalid: Bool, _ isLastByteInPacket: Bool, _ readingInvalidData: inout Data?) -> Message? {
        if byteIsInvalid {
            if readingInvalidData == nil {
                readingInvalidData = Data()
            }
            readingInvalidData?.append(byte)
        }

        if let invalidData = readingInvalidData,
           !byteIsInvalid || isLastByteInPacket {
            // We hit the end of a stretch of invalid data.
            readingInvalidData = nil
            return InvalidMessage(timeStamp: timeStamp, data: invalidData)
        }
        else {
            return nil
        }
    }

    @objc private func sysExTimedOut(_ timer: Timer) {
        sysExTimeOutTimer = nil
        if let message = finishSysExMessage(validEnd: false) {
            delegate?.parserDidReadMessages(self, messages: [message])
        }
    }

}
